Android/app/src/main/java/app/intra/socks/ReliabilityMonitor.java

```diff
@@ -0,0 +1,117 @@
+package app.intra.socks;
+
+import android.content.Context;
+import android.os.Bundle;
+import android.os.SystemClock;
+import app.intra.util.Names;
+import com.google.firebase.analytics.FirebaseAnalytics;
+import com.google.firebase.analytics.FirebaseAnalytics.Param;
+import java.net.Socket;
+import sockslib.server.io.Pipe;
+import sockslib.server.io.PipeListener;
+import sockslib.server.io.SocketPipe;
+import sockslib.server.listener.PipeInitializer;
+
+/**
+ * This is a pipe initializer that monitors post-connection socket reliability.
+ */
+public class ReliabilityMonitor implements PipeInitializer {
+  private final Context context;
+
+  ReliabilityMonitor(Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public Pipe initialize(Pipe pipe) {
+    pipe.addPipeListener(new ReliabilityListener(context));
+    return pipe;
+  }
+
+  /**
+   * This class converts pipe lifecycle callbacks into network reliability monitoring events.
+   */
+  private static class ReliabilityListener implements PipeListener {
+    private static final int HTTPS_PORT = 443;
+    private static final String RESET_MESSAGE = "Connection reset";
+    private final Context context;
+    private int downloadBytes = 0;
+    private int uploadBytes = 0;
+    private int uploadCount = 0;
+    private long startTime = -1;
+
+    ReliabilityListener(Context context) {
+      this.context = context;
+    }
+
+    private static int getPort(Pipe pipe) {
+      final Socket remoteSocket;
+      if (SocketPipe.INPUT_PIPE_NAME.equals(pipe.getName())) {
+        remoteSocket = (Socket) pipe.getAttribute(SocketPipe.ATTR_SOURCE_SOCKET);
+      } else if (SocketPipe.OUTPUT_PIPE_NAME.equals(pipe.getName())){
+        remoteSocket = (Socket) pipe.getAttribute(SocketPipe.ATTR_DESTINATION_SOCKET);
+      } else {
+        remoteSocket = null;
```

Comment:
You could return here, though I see the reason too keep the explicit check.

Reply:
Yeah, my thinking was that it would better to avoid two instances of `return -1`, i.e. the magic constant should only appear once.